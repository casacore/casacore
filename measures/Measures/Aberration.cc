//# Aberration.cc:  Aberration class
//# Copyright (C) 1995,1996,1997,1998,1999,2000,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casacore/measures/Measures/Aberration.h>
#include <casacore/casa/BasicSL/Constants.h>
#include <casacore/casa/Arrays/ArrayMath.h>
#include <casacore/measures/Measures/MeasTable.h>
#include <casacore/casa/System/AipsrcValue.h>

namespace casacore { //# NAMESPACE CASACORE - BEGIN

//# Constants
const Double Aberration::INTV = 0.04;

//# Static data
uInt Aberration::interval_reg = 0;
uInt Aberration::usejpl_reg = 0;

//# Constructors
Aberration::Aberration() : method(Aberration::STANDARD), lres(0) {
    fill();
}

Aberration::Aberration(const Aberration &other) {
    copy(other);
}

Aberration::Aberration(AberrationTypes type) :
method(type), lres(0) {
    fill();
}

Aberration &Aberration::operator=(const Aberration &other) {
    if (this != &other) {
	copy(other);
    }
    return *this;
}


void Aberration::init() {
    method = Aberration::STANDARD;
    fill();
}

void Aberration::init(AberrationTypes type) {
    method = type;
    fill();
}

void Aberration::copy(const Aberration &other) {
    method = other.method;
    checkEpoch = other.checkEpoch;
    for (Int i=0; i<3; i++) {
	aval[i] = other.aval[i];
	dval[i] = other.dval[i];
    }
    for (Int j=0; j<4; j++) {
	result[j] = other.result[j];
    }
}

//# Destructor
Aberration::~Aberration() {}

//# Operators
// Calculate Aberration 
const MVPosition &Aberration::operator()(Double epoch) {
    calcAber(epoch);
    Double dt = epoch - checkEpoch;
    Double fac = 1;
    if (AipsrcValue<Bool>::get(Aberration::usejpl_reg) && method != B1950) {
      fac /= MeasTable::Planetary(MeasTable::CAU);
    }
    lres++; lres %= 4;
    for (Int i=0; i<3; i++) {
      result[lres](i) = fac * (aval[i] + dt*dval[i]);
    }
    return result[lres];
}

//# Member functions

const MVPosition &Aberration::derivative(Double epoch) {
    calcAber(epoch);
    lres++; lres %= 4;
    Double fac = 1;
    if (AipsrcValue<Bool>::get(Aberration::usejpl_reg) && method != B1950) {
      fac /=  MeasTable::Planetary(MeasTable::CAU);
    }
    for (Int i=0; i<3; i++) {
      result[lres](i) = fac * dval[i];
    }
    return result[lres];
}

void Aberration::fill() {
  // Get the interpolation interval
  if (!Aberration::interval_reg) {
    interval_reg = 
      AipsrcValue<Double>::registerRC(String("measures.aberration.d_interval"),
				      Unit("d"), Unit("d"),
				      Aberration::INTV);
  }
  if (!Aberration::usejpl_reg) {
    usejpl_reg =
      AipsrcValue<Bool>::registerRC(String("measures.aberration.b_usejpl"),
				    False);
  }
  checkEpoch = 1e30;
}

void Aberration::refresh() {
    checkEpoch = 1e30;
}

void Aberration::calcAber(Double t) {
  if (!nearAbs(t, checkEpoch,
	       AipsrcValue<Double>::get(Aberration::interval_reg)) ||
      (AipsrcValue<Bool>::get(Aberration::usejpl_reg) &&
       method != B1950) ) {
    checkEpoch = t;
    switch (method) {
    case B1950:
      // Yes, this really should be the time in Julian centuries since January 	 
      // 0.5, 1900, not 1950.  And MJDB1900 should probably be named MJD1900 	 
      // since it is 15019.5, indicating a Julian instead of a Besselian (= 	 
      // tropical) date.
      t = (t - MeasData::MJDB1900)/MeasData::JDCEN;
      break;
    default:
      t = (t - MeasData::MJD2000)/MeasData::JDCEN;
      break;
    }
    Int i,j;
    Vector<Double> fa(13), dfa(13);
    for (i=0; i<3; i++) {
      aval[i] = dval[i] = Double(0);
    }
    Double dtmp, ddtmp, sdtmp, cdtmp;
    switch (method) {
    case B1950:
      {
        for (i=0; i<12; i++) {
          const Polynomial<Double>& aberArgP = MeasTable::aber1950Arg(i);
          fa(i) = aberArgP(t);
          dfa(i) = (aberArgP.derivative())(t);
        }
	CountedPtr<Matrix<Double> > mul = MeasTable::mulAber1950(t, 1e-6);
        DebugAssert (mul->contiguousStorage(), AipsError);
        const Double* mulAberV = mul->data();

        for (i=0; i<132; i++) {
          const Double* mulAberArgV = MeasTable::mulAber1950Arg(i);
          dtmp = ddtmp = 0; 
          for (j=0; j<12; j++) {
            dtmp += mulAberArgV[j] * fa(j);
            ddtmp += mulAberArgV[j] * dfa(j);
          }
          sdtmp = sin(dtmp);
          cdtmp = cos(dtmp);

          aval[0] += mulAberV[0] * sdtmp + mulAberV[1] * cdtmp;
          aval[1] += mulAberV[2] * sdtmp + mulAberV[3] * cdtmp;
          aval[2] += mulAberV[4] * sdtmp + mulAberV[5] * cdtmp;
          dval[0] += mulAberV[6] * sdtmp + mulAberV[7] * cdtmp +
            (mulAberV[0] * cdtmp - mulAberV[1] * sdtmp) * ddtmp;
          dval[1] += mulAberV[8] * sdtmp + mulAberV[9] * cdtmp +
            (mulAberV[2] * cdtmp - mulAberV[3] * sdtmp) * ddtmp;
          dval[2] += mulAberV[10] * sdtmp + mulAberV[11] * cdtmp +
            (mulAberV[4] * cdtmp - mulAberV[5] * sdtmp) * ddtmp;
          mulAberV += 12;
        }
        for (i=0; i<3; i++) {
          aval[i] /= C::c;
          dval[i] /= (C::c * MeasData::JDCEN);
        }
      }
      break;
      
    default:
      if (AipsrcValue<Bool>::get(Aberration::usejpl_reg)) {
	Vector<Double> mypl =
	  MeasTable::Planetary(MeasTable::EARTH, checkEpoch);
	for (i=0; i<3; i++) {
	  aval[i] = mypl[i + 3];
	  dval[i] = 0;
	}
      } else {
	for (i=0; i<13; i++) {
	  const Polynomial<Double>& aberArgP = MeasTable::aberArg(i);

	  fa(i) = aberArgP(t);
	  dfa(i) = (aberArgP.derivative())(t);
	}
	CountedPtr<Matrix<Double> > mul = MeasTable::mulAber(t, 1e-6);
        DebugAssert (mul->contiguousStorage(), AipsError);
        const Double* mulAberV = mul->data();
	for (i=0; i<80; i++) {
	  const Double* mulAberArgV = MeasTable::mulAberArg(i);
	  dtmp = ddtmp = 0; 
	  for (j=0; j<6; j++) {
	    dtmp  += mulAberArgV[j] * fa[j];
	    ddtmp += mulAberArgV[j] * dfa[j];
	  }
	  sdtmp = sin(dtmp);
	  cdtmp = cos(dtmp);
	  
	  aval[0] += mulAberV[0] * sdtmp + mulAberV[1] * cdtmp;
	  aval[1] += mulAberV[2] * sdtmp + mulAberV[3] * cdtmp;
	  aval[2] += mulAberV[4] * sdtmp + mulAberV[5] * cdtmp;
	  dval[0] += mulAberV[6] * sdtmp + mulAberV[7] * cdtmp +
	    (mulAberV[0] * cdtmp - mulAberV[1] * sdtmp) * ddtmp;
	  dval[1] += mulAberV[8] * sdtmp + mulAberV[9] * cdtmp +
	    (mulAberV[2] * cdtmp - mulAberV[3] * sdtmp) * ddtmp;
	  dval[2] += mulAberV[10] * sdtmp + mulAberV[11] * cdtmp +
	    (mulAberV[4] * cdtmp - mulAberV[5] * sdtmp) * ddtmp;
          mulAberV += 12;
	}
	for (i=0; i<17; i++) {
	  const Double* mulAberArgV = MeasTable::mulAberSunArg(i);

	  dtmp = ddtmp = 0;
	  for (j=0; j<7; j++) {
	    dtmp  += mulAberArgV[j] * fa[j + 1];
	    ddtmp += mulAberArgV[j] * dfa[j + 1];
	  }
	  
	  sdtmp = sin(dtmp);
	  cdtmp = cos(dtmp);

	  const Vector<Double>& mulAberV = MeasTable::mulSunAber(i);

	  aval[0] += mulAberV[0] * sdtmp + mulAberV[1] * cdtmp;
	  aval[1] += mulAberV[2] * sdtmp + mulAberV[3] * cdtmp;
	  aval[2] += mulAberV[4] * sdtmp + mulAberV[5] * cdtmp;
	  dval[0] += (mulAberV[0] * cdtmp - mulAberV[1] * sdtmp) * ddtmp;
	  dval[1] += (mulAberV[2] * cdtmp - mulAberV[3] * sdtmp) * ddtmp;
	  dval[2] += (mulAberV[4] * cdtmp - mulAberV[5] * sdtmp) * ddtmp;
	}
	for (i=0; i<17; i++) {
	  const Double* mulAberArgV = MeasTable::mulAberEarthArg(i);

	  dtmp = ddtmp = 0;
	  for (j=0; j<5; j++) {
	    dtmp  += mulAberArgV[j] *  fa[j + 8];
	    ddtmp += mulAberArgV[j] * dfa[j + 8];
	  }

	  sdtmp = sin(dtmp);
	  cdtmp = cos(dtmp);
	          
	  const Vector<Double>& mulAberV = MeasTable::mulEarthAber(i);

	  aval[0] += mulAberV[0] * sdtmp;
	  aval[1] += mulAberV[1] * cdtmp;
	  aval[2] += mulAberV[2] * cdtmp;
	  dval[0] += mulAberV[0] * cdtmp * ddtmp;
	  dval[1] += -mulAberV[1] * sdtmp * ddtmp;
	  dval[2] += -mulAberV[2] * sdtmp * ddtmp;
	}
	for (i=0; i<3; i++) {
	  aval[i] /= C::c;
	  dval[i] /= (C::c * MeasData::JDCEN);
	}
      }
      break;
    }
  }
}

} //# NAMESPACE CASACORE - END

