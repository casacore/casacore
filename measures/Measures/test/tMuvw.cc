//# tMuvw.cc: This program tests Muvw class
//# Copyright (C) 1998-2000,2002,2007
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This program is free software; you can redistribute it and/or modify it
//# under the terms of the GNU General Public License as published by the Free
//# Software Foundation; either version 2 of the License, or (at your option)
//# any later version.
//#
//# This program is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
//# more details.
//#
//# You should have received a copy of the GNU General Public License along
//# with this program; if not, write to the Free Software Foundation, Inc.,
//# 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

//# Includes
#include <casacore/casa/aips.h>
#include <casacore/casa/Exceptions/Error.h>
#include <casacore/measures/Measures.h>
#include <casacore/measures/Measures/Aberration.h>
#include <casacore/casa/Quanta/RotMatrix.h>
#include <casacore/casa/Quanta/Euler.h>
#include <casacore/measures/Measures/MEpoch.h>
#include <casacore/measures/Measures/MDirection.h>
#include <casacore/casa/Quanta/MVDirection.h>
#include <casacore/casa/Quanta/MVBaseline.h>
#include <casacore/measures/Measures/MPosition.h>
#include <casacore/measures/Measures/MCuvw.h>
#include <casacore/measures/Measures/Muvw.h>
#include <casacore/casa/Quanta/MVuvw.h>
#include <casacore/casa/Arrays/ArrayMath.h>
#include <casacore/casa/Arrays/ArrayLogical.h>
#include <casacore/casa/iostream.h>

#include <casacore/casa/namespace.h>
int main()
{
    try {
      	cout << "Test measure class Muvw" << endl;
	cout << "--------------------------------------" << endl;

	cout << endl << "Muvw state transition matrix:\n" << endl;
	cout << MCuvw::showState() << endl;

	MEpoch tbm(Quantity(50927.92931, "d"));
	MPosition pos(MVPosition(-4750915.84032, 2792906.17778, 
				 -3200483.75028), 
		      MPosition::ITRF);
	MeasFrame mf(tbm, pos);
	mf.set(MDirection(Quantity(10, "deg"), Quantity(80, "deg"),
			  MDirection::HADEC));
	MVuvw mvb0(100 ,10, 0);
	cout << "uvw: " << mvb0 << endl;

	Muvw::Ref mbref0(Muvw::ITRF, mf);
	Muvw mb0(mvb0, mbref0);

	cout << "uvw: " << mb0 << endl;
	cout << "uvw reference: " << mbref0 << endl;

	Muvw::Ref mbref1(Muvw::J2000);

	cout << "uvw reference: " << mbref1 << endl;

	cout << "Test uvw conversion ..." << endl;
	cout << "--------------------------------------" << endl;
	Muvw::Convert bconv(mb0, mbref1);
	cout << "Converted " << mb0 << endl <<
	  " to " << mbref1 << endl <<
	  " as " << bconv() << endl;

	Muvw::Convert bconvb(mbref1, mbref0);
	if (allNearAbs(mb0.getValue().getValue(),
		       bconvb(bconv()).getValue().getValue(),
		       1e-8)) {
	  cout << "Back      " << mb0 << " : ok" << endl;
	} else {
	  cout << "Back      " << mb0 << " : not ok" << endl <<
	  " as " << bconvb(bconv()) << endl;
	};

	cout << "--------------------------------------" << endl;
	cout << "Make uvw from Baseline" << endl;
	MVBaseline mvb1(100, 10 ,1);
	MVDirection mvd(Quantity(30, "deg"), Quantity(45, "deg"));
	MVuvw mvu0(mvb1, mvd);
	cout << "Baseline: " << mvb1 << mvb1.getAngle("deg") << endl;
	cout << "Direction:" << mvd << mvd.getAngle("deg") << endl;
	cout << "uvw:      " << mvu0 << mvu0.getAngle("deg") << endl;

	cout << "--------------------------------------" << endl;
	cout << "Testing all conversions forward/backward" << endl;

	Vector<Double> tvec(3);
	tvec = 0.0;
	Bool isok = True;
	for (uInt i=Muvw::J2000; i<Muvw::N_Types; i++) {
	  for (uInt j=Muvw::J2000; j<Muvw::N_Types; j++) {
	    Muvw::Ref rin(i, mf);
	    Muvw::Ref rout(j, mf);
	    Muvw mb0(mvb0, rin);
	    Muvw::Convert forw(rin, rout);
	    Muvw::Convert backw(rout, rin);
	    if (!allNearAbs(mb0.getValue().getValue() -
			 backw(forw(mb0)).getValue().getValue(), 
			 tvec, 1e-4)) {
	      cout << Muvw::showType(i) << " to " <<
		Muvw::showType(j) << ": " <<
		mb0.getValue().getValue() -
		backw(forw(mb0)).getValue().getValue() << endl;
	      isok = False;
	    };
	  };
	};
      if (isok) {
	cout << "All forward/backward uvw conversions: ok" << endl;
      } else {
	cout << "Some forward/backward uvw conversions wrong" << endl;
      };
	cout << "--------------------------------------" << endl;

	cout << "Exercise all MVuvw functions" << endl;
	{
	  MVuvw x(mvb0);
	  if (x != mvb0) cout << "Copy constructor error" << endl;
	  x = mvb0;
	  if (x != mvb0) cout << "Assignment error" << endl;
	  Vector<Quantum<Double> > vq(3);
	  vq = Quantity(23, "m");
	  x.putValue(vq);
	  cout << "putValue:       " << vq << ", " << x << endl;
	  cout << "uvwAngle:  " << x.uvwAngle(mvb0) << endl;
	  cout << "uvwAngle:  " << x.uvwAngle(mvb0, "deg") << endl;
	  cout << "get:            " << x.get() << endl;
	  cout << "getRecordValue: " << x.getRecordValue() << endl;
	  cout << "separation:     " << x.separation(mvb0) << endl;
	  cout << "separation:     " << x.separation(mvb0, "deg") << endl;
	  cout << "crossProduct:   " << x.crossProduct(mvb0) << endl;
	  cout << "getAngle:       " << x.getAngle() << endl;
	  cout << "getAngle:       " << x.getAngle("deg") << endl;
	  cout << "getlength:      " << x.getLength("cm") << endl;
	  cout << "radius:         " << x.radius() << endl;
	  cout << "getXRecordValue:" << x.getXRecordValue() << endl;
	  Vector<Double> x1(3);
	  x1(0) = 30;
	  x1(1) = 40;
	  x1(2) = 0;
	  x.putVector(x1);
	  cout << "putVector:      " << x1 << ", " << x << endl;
	  MVuvw x2(vq);
	  cout << "VQ constructor: " << x2 << endl;
	  cout << "Pos constructor:" << MVuvw(x, x2) << endl;
	  cout << "Q constructor:  " << MVuvw(Quantity(50, "m")) << endl;
	  cout << "QV constructor: " << MVuvw(x2.getAngle()) << endl;
	  cout << "QV constructor: " << MVuvw(Quantity(34,"m"),
						   x2.getAngle()) << endl;
	  cout << "V constructor:  " << MVuvw(x1) << endl;
	  cout << "D constructor:  " << MVuvw(Double(78)) << endl;
	  cout << "operator+:      " << x+x2 << endl;
	  cout << "operator-:      " << x-x2 << endl;
	  cout << "operator-pre-:  " << -x2 << endl;
	  RotMatrix rm(Euler(25, 1, 0, 0));
	  cout << "operator*:      " << x2*rm << endl;
	  cout << "operator*:      " << x2*2 << endl;
	  MVuvw::assure(x);
	  cout << "assure:         " << "ok" << endl;
	  cout << "getLength:      " << x.getLength() << endl;
	  cout << "operator*:      " << x*x1 << endl;
	  cout << "operator*       " << x*x2 << endl;
	  cout << "operator*:      " << x1*x << endl;
	  MeasValue *xmvu = x.clone();
	  cout << "clone:          " << *xmvu << endl;
	  delete xmvu;
	  cout << "getVector:      " << x.getVector() << endl;
	  cout << "near:           " << x.near(x2) << endl;
	  cout << "near:           " << x.near(x2, Quantity(1, "deg")) << endl;
	  cout << "nearAbs:        " << x.nearAbs(x2) << endl;
	  cout << "!=:             " << (x != x2) << endl;
	  cout << "==:             " << (x == x2) << endl;
	  cout << "Original:       " << x << endl;
	  Double xa;
	  x.adjust(xa);
	  cout << "adjust:         " << x << endl;
	  x.readjust(xa);
	  cout << "readjust:       " << x << endl;

	  cout << "All MVuvw functions: ok" << endl;
	  cout << "----------------------------" << endl;
	}

 	cout << "Exercise all Muvw function" << endl;
	{
	  Muvw mb(mvb0, Muvw::B1950);
	  String s0("azel");
	  Muvw::Types tp;
	  Muvw::Ref mr;
	  cout << "getType:        " << Muvw::getType(tp, s0) << ", ";
	  // next () to stop egcs warning
	  cout << (uInt)tp << endl;
	  cout << "giveMe:         " << mb.giveMe(mr, s0) << ", ";
	  cout << mr << endl;
	  cout << "setRefString:   " << mb.setRefString("hadec") << ", ";
	  cout << mb << endl;
	  Muvw::assure(mb);
	  cout << "assure:         " << "ok" << endl;
	  Measure *xmu = mb.clone();
	  cout << "clone:          " << *xmu << endl;
	  delete xmu;
	  cout << "get:            " << mb.get("cm") << endl;
	  cout << "getAngle:       " << mb.getAngle("deg") << endl;
	  cout << "getDefaultType: " << mb.getDefaultType() << endl;
	  cout << "getRefString:   " << mb.getRefString() << endl;
	  cout << "Original:       " << mb << endl;
	  Muvw cpc(mb);
	  cout << "Ctor copy:      " << cpc << endl;
	  Muvw cpas;
	  cpas = mb;
	  cout << "Assign:         " << cpas << endl;
	  cout << "tellMe:         " << mb.tellMe() << endl;
	  cout << "showMe:         " << mb.showMe() << endl;

	  cout << "All Muvw functions: ok" << endl;
	  cout << "---------------------------" << endl;
	}

    } catch (std::exception& x) {
	cout << x.what() << endl;
    } 

    try {
      cout << "Test Muvw exception" << endl;
      cout << "---------------------------" << endl;
      MEpoch x;
      Muvw::assure(x);

    } catch (std::exception& x) {
	cout << x.what() << endl;
    } 

    try {
      cout << "Test MVuvw exception" << endl;
      cout << "---------------------------" << endl;
      MVEpoch x;
      MVuvw::assure(x);

    } catch (std::exception& x) {
	cout << x.what() << endl;
    } 
    cout << "---------------------------" << endl;

    return 0;
}
