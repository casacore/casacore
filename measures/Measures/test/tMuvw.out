Test measure class Muvw
--------------------------------------

Muvw state transition matrix:

   |  0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21
----------------------------------------------------------------------
  0| -- 10 10 18  4  5  4  4  2 18 18 18 18 18 18 34 10 10  2 18 18 47
   |     1  1 14  4  5  4  4  8 14 14 14 14 14 14 15  1  1  8 14 14 21
  1| 12 -- 13 12 12 12 12 12 12 12 12 12 12 12 12 12 36 13 12 12 12 12
   |  0     2  0  0  0  0  0  0  0  0  0  0  0  0  0 16  2  0  0  0  0
  2| 16 16 -- 16 16 16 16 16 16 16 16 16 16 16 16 16 16 38 16 16 16 16
   |  1  1     1  1  1  1  1  1  1  1  1  1  1  1  1  1 17  1  1  1  1
  3| 32 32 32 -- 21 21 21 21 32 22 22 22 22 22 32 32 32 32 32 22 22 32
   | 14 14 14     4  4  4  4 14 20 20 20 20 20 14 14 14 14 14 20 20 14
  4|  6  6  6 20 --  8 11 11  3 20 20 20 20 20  6  6  6  6  3 20 20  6
   |  0  0  0  3     5  6  6  8  3  3  3  3  3  0  0  0  0  8  3  3  0
  5|  7  7  7  9  9 --  9  9  7  9  9  9  9  9  7  7  7  7  7  9  9  7
   |  0  0  0  4  4     4  4  0  4  4  4  4  4  0  0  0  0  0  4  4  0
  6| 14 14 14 14 14 14 -- 15 14 14 14 14 14 14 14 14 14 14 14 14 14 14
   |  4  4  4  4  4  4     7  4  4  4  4  4  4  4  4  4  4  4  4  4  4
  7| 17 17 17 17 17 17 17 -- 17 17 17 17 17 17 17 17 17 17 17 17 17 17
   |  6  6  6  6  6  6  6     6  6  6  6  6  6  6  6  6  6  6  6  6  6
  8|  0  0  0  0  1  0  1  1 --  0  0  0  0  0  0  0  0  0 40  0  0  0
   |  0  0  0  0  4  0  4  4     0  0  0  0  0  0  0  0  0 18  0  0  0
  9| 27 27 27 27 27 27 27 27 27 -- 23 23 24 24 27 27 27 27 27 43 27 27
   | 20 20 20 20 20 20 20 20 20    10 10 12 12 20 20 20 20 20 19 20 20
 10| 25 25 25 25 25 25 25 25 25 25 -- 28 25 25 25 25 25 25 25 25 25 25
   |  9  9  9  9  9  9  9  9  9  9    11  9  9  9  9  9  9  9  9  9  9
 11| 30 30 30 30 30 30 30 30 30 30 30 -- 30 30 30 30 30 30 30 30 30 30
   | 10 10 10 10 10 10 10 10 10 10 10    10 10 10 10 10 10 10 10 10 10
 12| 26 26 26 26 26 26 26 26 26 26 26 26 -- 29 26 26 26 26 26 26 26 26
   |  9  9  9  9  9  9  9  9  9  9  9  9    13  9  9  9  9  9  9  9  9
 13| 31 31 31 31 31 31 31 31 31 31 31 31 31 -- 31 31 31 31 31 31 31 31
   | 12 12 12 12 12 12 12 12 12 12 12 12 12    12 12 12 12 12 12 12 12
 14| 19 19 19 33 19 19 19 19 19 33 33 33 33 33 -- 19 19 19 19 33 33 19
   |  0  0  0  3  0  0  0  0  0  3  3  3  3  3     0  0  0  0  3  3  0
 15| 35 35 35 35 35 35 35 35 35 35 35 35 35 35 35 -- 35 35 35 35 35 35
   |  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0     0  0  0  0  0  0
 16| 37 37 37 37 37 37 37 37 37 37 37 37 37 37 37 37 -- 37 37 37 37 37
   |  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1     1  1  1  1  1
 17| 39 39 39 39 39 39 39 39 39 39 39 39 39 39 39 39 39 -- 39 39 39 39
   |  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2  2     2  2  2  2
 18| 41 41 41 41 41 41 41 41 41 41 41 41 41 41 41 41 41 41 -- 41 41 41
   |  8  8  8  8  8  8  8  8  8  8  8  8  8  8  8  8  8  8     8  8  8
 19| 42 42 42 42 42 42 42 42 42 42 42 42 42 42 42 42 42 42 42 -- 42 42
   |  9  9  9  9  9  9  9  9  9  9  9  9  9  9  9  9  9  9  9     9  9
 20| 45 45 45 45 45 45 45 45 45 44 44 44 44 44 45 45 45 45 45 44 -- 45
   |  3  3  3  3  3  3  3  3  3  9  9  9  9  9  3  3  3  3  3  9     3
 21| 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 46 --
   |  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0   

uvw: [100, 10, 0]
uvw: uvw: [100, 10, 0]
uvw reference: Reference for an uvw with Type: ITRF,
Frame: Epoch: 50927::22:18:12.3840 (TDB = 50927.9, UT1 = 50927.9, TT = 50927.9)
       Position: [-4.75092e+06, 2.79291e+06, -3.20048e+06]
        (Longitude = 2.61014 Latitude = -0.526138)
       Direction: [0.17101, 0.0301537, 0.984808]
        (J2000 = [-33.1687, 80.0127] deg)
uvw reference: Reference for an uvw with Type: J2000
Test uvw conversion ...
--------------------------------------
Converted uvw: [100, 10, 0]
 to Reference for an uvw with Type: J2000
 as uvw: [99.9944, 10.0562, -1.50948e-07]
Back      uvw: [100, 10, 0] : not ok
 as uvw: [100, 10, -1.50877e-07]
--------------------------------------
Make uvw from Baseline
Baseline: [100, 10, 1][5.71059, 0.570096] deg
Direction:[0.612372, 0.353553, 0.707107][30, 45] deg
uvw:      [-41.3397, -64.0657, 65.4799][-122.833, 40.6561] deg
--------------------------------------
Testing all conversions forward/backward
All forward/backward uvw conversions: ok
--------------------------------------
Exercise all MVuvw functions
putValue:       [23 m, 23 m, 23 m], [23, 23, 23]
uvwAngle:  -2.1853
uvwAngle:  -125.209 deg
get:            [39.8372, 0.785398, 0.61548]
getRecordValue: [0.785398 rad, 0.61548 rad, 39.8372 m]
separation:     3.14159
separation:     180 deg
crossProduct:   [-230, 2300, -2070]
getAngle:       [0.785398, 0.61548] rad
getAngle:       [45, 35.2644] deg
getlength:      3983.72 cm
radius:         39.8372
getXRecordValue:[23 m, 23 m, 23 m]
putVector:      [30, 40, 0], [30, 40, 0]
VQ constructor: [23, 23, 23]
Pos constructor:[7.07107, -28.5774, 40.4145]
Q constructor:  [0, 0, 50]
QV constructor: [0.57735, 0.57735, 0.57735]
QV constructor: [19.6299, 19.6299, 19.6299]
V constructor:  [30, 40, 0]
D constructor:  [0, 0, 78]
operator+:      [53, 63, 23]
operator-:      [7, 17, -23]
operator-pre-:  [-23, -23, -23]
operator*:      [23, 19.7536, 25.8418]
operator*:      [46, 46, 46]
assure:         ok
getLength:      50 m
operator*:      2500
operator*       1610
operator*:      2500
clone:          [30, 40, 0]
getVector:      [30, 40, 0]
near:           0
near:           0
nearAbs:        0
!=:             1
==:             0
Original:       [30, 40, 0]
adjust:         [0.6, 0.8, 0]
readjust:       [30, 40, 0]
All MVuvw functions: ok
----------------------------
Exercise all Muvw function
getType:        1, 10
giveMe:         1, Reference for an uvw with Type: AZEL
setRefString:   1, uvw: [100, 10, 0]
assure:         ok
clone:          uvw: [100, 10, 0]
get:            [10000, 1000, 0] cm
getAngle:       [5.71059, 0] deg
getDefaultType: ITRF
getRefString:   HADEC
Original:       uvw: [100, 10, 0]
Ctor copy:      uvw: [100, 10, 0]
Assign:         uvw: [100, 10, 0]
tellMe:         uvw
showMe:         uvw
All Muvw functions: ok
---------------------------
Test Muvw exception
---------------------------
Illegal Measure type argument: uvw
Test MVuvw exception
---------------------------
Illegal MeasValue type argument: MVuvw
---------------------------
